official/recommendation/neumf_model.py

```diff
@@ -0,0 +1,182 @@
+
+    self.num_users = num_users
+    self.num_items = num_items
+    self.mf_dim = mf_dim
+    self.model_layers = model_layers
+    self.learning_rate = learning_rate
+
+  def __call__(self, multi_gpu, batch_size):
+    """Create a NeuMF recommendation model.
+
+    Args:
+      multi_gpu: A boolean flag for multi_gpu option
+      batch_size: The number of examples processed in each training batch.
+    Returns:
+      neumf_model: An instance of tf.keras Model as the NeuMF model.
+    """
+    # Input variables
+    user_input = tf.keras.layers.Input(
+        shape=(1,), dtype='int32', name='user_input')
+    item_input = tf.keras.layers.Input(
+        shape=(1,), dtype='int32', name='item_input')
+
+    # Initializer for embedding layer
+    embedding_initializer = tf.keras.initializers.RandomNormal(stddev=0.01)
+    # Embedding layer of GMF and MLP
+    mf_embedding_user = tf.keras.layers.Embedding(
+        self.num_users,
+        self.mf_dim,
+        embeddings_initializer=embedding_initializer,
+        input_length=1)
+    mf_embedding_item = tf.keras.layers.Embedding(
+        self.num_items,
+        self.mf_dim,
+        embeddings_initializer=embedding_initializer,
+        input_length=1)
+
+    mlp_embedding_user = tf.keras.layers.Embedding(
+        self.num_users,
+        self.model_layers[0]//2,
+        embeddings_initializer=embedding_initializer,
+        input_length=1)
+    mlp_embedding_item = tf.keras.layers.Embedding(
+        self.num_items,
+        self.model_layers[0]//2,
+        embeddings_initializer=embedding_initializer,
+        input_length=1)
+
+    # GMF part
+    mf_user_latent = tf.keras.layers.Flatten()(mf_embedding_user(user_input))
+    mf_item_latent = tf.keras.layers.Flatten()(mf_embedding_item(item_input))
+    # Element-wise multiply
+    mf_vector = tf.keras.layers.multiply([mf_user_latent, mf_item_latent])
+
+    # MLP part
+    mlp_user_latent = tf.keras.layers.Flatten()(mlp_embedding_user(user_input))
+    mlp_item_latent = tf.keras.layers.Flatten()(mlp_embedding_item(item_input))
+    # Concatenation of two latent features
+    mlp_vector = tf.keras.layers.concatenate([mlp_user_latent, mlp_item_latent])
+
+    num_layer = len(self.model_layers)  # Number of layers in the MLP
+    for idx in xrange(1, num_layer):
+      model_layer = tf.keras.layers.Dense(
+          self.model_layers[idx],
+          activation='relu'
+      )
+      mlp_vector = model_layer(mlp_vector)
+
+    # Concatenate GMF and MLP parts
+    predict_vector = tf.keras.layers.concatenate([mf_vector, mlp_vector])
+
+    # Final prediction layer
+    prediction = tf.keras.layers.Dense(
+        1, activation='sigmoid', kernel_initializer='lecun_uniform',
+        name='prediction')(predict_vector)
+
+    # Instantiate a tf.keras model
+    model = tf.keras.models.Model(
+        inputs=[user_input, item_input], outputs=prediction)
+
+    # Use multiple gpus
+    if multi_gpu:
```

Comment:
Will update this multi_gpu mode with distribution strategy, based on discussions with @robieta